batch2/batch/batch.py

```diff
@@ -1,19 +1,1278 @@
+        log.info(f'job {job.id} mark complete')
+        await job.mark_complete(pod)
+        return
+
+    if pod and pod.status and pod.status.phase == 'Unknown':
+        log.info(f'job {job.id} mark unscheduled -- pod phase is unknown')
+        await job.mark_unscheduled()
+        return
+
+
+class DeblockedIterator:
+    def __init__(self, it):
+        self.it = it
+
+    def __aiter__(self):
+        return self
+
+    def __anext__(self):
+        return blocking_to_async(app['blocking_pool'], self.it.__next__)
+
+
+async def pod_changed(pod):
+    job = await Job.from_pod(pod)
+    await update_job_with_pod(job, pod)
+
+
+async def refresh_pods():
+    log.info(f'refreshing pods')
+
+    pod_jobs = [Job.from_record(record) for record in await db.jobs.get_records_where({'state': 'Running'})]
+
+    pods = app['driver'].list_pods()
+    log.info(f'batch had {len(pods)} pods')
+
+    seen_pods = set()
+    for pod_dict in pods:
+        pod = v1.api_client._ApiClient__deserialize(pod_dict, kube.client.V1Pod)
+        pod_name = pod.metadata.name
+        seen_pods.add(pod_name)
+        asyncio.ensure_future(pod_changed(pod))
+
+    if len(seen_pods) != len(pod_jobs):
+        log.info('restarting running jobs with pods not seen in batch')
+
+    async def restart_job(job):
+        log.info(f'restarting job {job.id}')
+        await update_job_with_pod(job, None)
+    asyncio.gather(*[restart_job(job)
+                     for job in pod_jobs
+                     if job._pod_name not in seen_pods])
+
+
+async def driver_event_loop():
+    await asyncio.sleep(1)
+    while True:
+        try:
+            object = await app['driver'].complete_queue.get()
+            pod = v1.api_client._ApiClient__deserialize(object, kube.client.V1Pod)
+            await pod_changed(pod)
+        except Exception as exc:
+            log.exception(f'driver event loop failed due to exception: {exc}')
+
+
+async def polling_event_loop():
+    await asyncio.sleep(1)
+    while True:
+        try:
+            await refresh_pods()
+        except Exception as exc:
+            log.exception(f'polling event loop failed due to exception: {exc}')
+        await asyncio.sleep(60 * 10)
+
+
+async def db_cleanup_event_loop():
+    await asyncio.sleep(1)
+    while True:
+        try:
+            for record in await db.batch.get_finished_deleted_records():
+                batch = Batch.from_record(record, deleted=True)
+                await batch.delete()
+        except Exception as exc:  # pylint: disable=W0703
+            log.exception(f'Could not delete batches due to exception: {exc}')
+        await asyncio.sleep(REFRESH_INTERVAL_IN_SECONDS)
+
+
+async def profile_loop():
+    await asyncio.sleep(1)
+    while True:
+        app['profile'].disable()
+        app['profile'].print_stats(sort='time')
+        app['profile'].enable()
+        await asyncio.sleep(15)
+
+
+@routes.post('/api/v1alpha/instances/activate')
+# @rest_authenticated_users_only
```

Comment:
Yeah, these are going to need to get protected.  I think we should have a batch secret token (instance token?) that is passed to instances when they are created and included in an Authorization header in the usual way and verified by batch2.